src/filesystem.cc

```diff
@@ -1427,32 +1457,36 @@ S3FileSystem::CleanPath(const std::string& s3_path, std::string* clean_path)
 }
 
 S3FileSystem::S3FileSystem(
-    const Aws::SDKOptions& options, const std::string& s3_path)
-    : options_(options),
-      s3_regex_(
+    const std::string& s3_path, const std::string& secret_key,
+    const std::string& key_id, const std::string& region,
+    const std::string& session_token, const std::string& profile_name)
+    : s3_regex_(
           "s3://(http://|https://|)([0-9a-zA-Z\\-.]+):([0-9]+)/"
           "([0-9a-z.\\-]+)(((/[0-9a-zA-Z.\\-_]+)*)?)")
 {
+  // init aws api if not already
+  init_mu_.lock();
+  if (!init_count_++) {
+    Aws::InitAPI(options_);
+  }
+  init_mu_.unlock();
+
```

Comment:
```suggestion
{
  // init aws api if not already
  std::lock_guard<std::mutex> lk(init_mu_);
  if (!init_count_++) {
    Aws::InitAPI(options_);
  }
}
```

Reply:
I think we should revert to  `std::call_once`, the Init / Shutdown seems to be a one-time call in respect to the process? Should isolate it from possible multiple S3 clients for different credentials

Reply:
Yes. Updated.